core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java

```diff
@@ -80,20 +91,32 @@ public FilterOnExcludedValuesRecordCursorFactory(
         this.filter = filter;
         cursorFactoriesIdx = new int[]{0};
         cursorFactories = new ObjList<>(nKeyValues);
-        if (orderByMnemonic == OrderByMnemonic.ORDER_BY_INVARIANT) {
+        if (orderByMnemonic == OrderByMnemonic.ORDER_BY_INVARIANT && !orderByTimestamp) {
             cursor = new DataFrameRecordCursorImpl(new SequentialRowCursorFactory(cursorFactories, cursorFactoriesIdx), false, filter, columnIndexes);
         } else {
             cursor = new DataFrameRecordCursorImpl(new HeapRowCursorFactory(cursorFactories, cursorFactoriesIdx), false, filter, columnIndexes);
         }
-        this.followedOrderByAdvice = followedOrderByAdvice;
+        this.followedOrderByAdvice = orderByKeyColumn || orderByTimestamp;
         this.columnIndexes = columnIndexes;
+
+        comparator = this::compareStrFunctions;
+        comparatorDesc = this::compareStrFunctionsDesc;
     }
 
     @Override
     public boolean followedOrderByAdvice() {
         return followedOrderByAdvice;
     }
 
+    @Override
+    public int getScanDirection() {
+        if (dataFrameCursorFactory.getOrder() == DataFrameCursorFactory.ORDER_ASC
+                && cursor.getRowCursorFactory() instanceof HeapRowCursorFactory) {
```

Comment:
nit: would be good to avoid `instanceof` here and instead set variables in the constructor for example. There is a growing theory that `instanceof` can cause expensive deoptmisations

Reply:
Fixed in f52f44b7